Sources/BrowserServicesKit/Suggestions/SuggestionLoading.swift

```diff
@@ -21,141 +21,92 @@ import Foundation
+        self.processing = SuggestionProcessing(urlFactory: urlFactory)
     }
 
     public func getSuggestions(query: Query,
-                               maximum: Int,
-                               completion: @escaping ([Suggestion]?, Error?) -> Void) {
+                               completion: @escaping (SuggestionResult?, Error?) -> Void) {
         guard let dataSource = dataSource else {
             completion(nil, SuggestionLoaderError.noDataSource)
             return
         }
 
         if query.isEmpty {
-            completion([], nil)
+            completion(.empty, nil)
             return
         }
 
+        // 1) Getting all necessary data
         let bookmarks = dataSource.bookmarks(for: self)
-        var bookmarkSuggestions: [Suggestion]!
-        var remoteSuggestions: [Suggestion]?
-        var remoteSuggestionsError: Error?
-        let group = DispatchGroup()
-
-        group.enter()
-        DispatchQueue.global().async {
-            // Bookmark suggestions
-            bookmarkSuggestions = self.bookmarkSuggestions(from: bookmarks, for: query)
-            group.leave()
-        }
+        let history = dataSource.history(for: self)
+        var apiResult: APIResult?
+        var apiError: Error?
 
-        // Remote suggestions
+        let group = DispatchGroup()
         group.enter()
         dataSource.suggestionLoading(self,
                                      suggestionDataFromUrl: Self.remoteSuggestionsUrl,
-                                     withParameters: [Self.searchParameter: query]) { [weak self] data, error in
+                                     withParameters: [Self.searchParameter: query]) { data, error in
             defer { group.leave() }
-            guard let self = self, let data = data else {
-                remoteSuggestionsError = error
+            guard let data = data else {
+                apiError = error
                 return
             }
-
-            do {
-                remoteSuggestions = try self.remoteSuggestions(from: data)
-            } catch {
-                remoteSuggestionsError = error
+            guard let result = try? JSONDecoder().decode(APIResult.self, from: data) else {
+                apiError = SuggestionLoaderError.parsingFailed
+                return
             }
+            apiResult = result
         }
 
-        group.notify(queue: .global()) {
-            let result = Self.result(maximum: maximum,
-                                     bookmarkSuggestions: bookmarkSuggestions,
-                                     remoteSuggestions: remoteSuggestions ?? [])
-
-            DispatchQueue.main.async {
-                guard !result.isEmpty || remoteSuggestionsError == nil else {
-                    completion(nil, SuggestionLoaderError.failedToObtainData)
-                    return
+        // 2) Processing it
+        group.notify(queue: .main) { [weak self] in
+            DispatchQueue.global(qos: .userInitiated).async {
```

Comment:
Why do we need it to synchronize over the main queue here, what was wrong with direct `.notify` on the global queue?

Reply:
👍 true. I probably forgot to remove it with other stuff that was happening before `DispatchQueue.global(qos: .userInitiated).async`